madmom/ml/hmm.pyx

```diff
@@ -518,52 +550,61 @@ class HiddenMarkovModel(object):
         """
         # transition model stuff
         tm = self.transition_model
-        cdef unsigned int [::1] tm_states = tm.states
-        cdef unsigned int [::1] tm_pointers = tm.pointers
+        cdef uint32_t [::1] tm_states = tm.states
+        cdef uint32_t [::1] tm_pointers = tm.pointers
         cdef double [::1] tm_probabilities = tm.probabilities
-        cdef unsigned int num_states = tm.num_states
+        cdef int num_states = tm.num_states
 
         # observation model stuff
         om = self.observation_model
-        cdef unsigned int num_observations = len(observations)
-        cdef unsigned int [::1] om_pointers = om.pointers
-        cdef double [:, ::1] om_densities = om.densities(observations)
+        cdef uint32_t [::1] om_pointers = om.pointers
+        # TODO: check why ascontiguousarray is needed for framewise processing
+        # make sure we can iterate over the observations
+        cdef double [:, ::1] om_densities = np.ascontiguousarray(
+            np.atleast_2d(om.densities(observations)))
+        cdef int num_observations = len(om_densities)
+
+        # reset HMM
+        if reset:
+            self.reset()
 
         # forward variables
-        cdef double[:, ::1] fwd = np.zeros((num_observations + 1, num_states),
-                                           dtype=np.float)
+        cdef double[::1] fwd_prev = self._prev
+        cdef double[:, ::1] fwd_cur = \
+            np.zeros((num_observations, num_states), dtype=np.float)
+
         # define counters etc.
-        cdef unsigned int prev_pointer, frame, state, cur, prev
+        cdef int prev_pointer, frame, state
         cdef double prob_sum, norm_factor
 
-        # init forward variables
-        for state in range(self.initial_distribution.shape[0]):
-            fwd[0, state] = self.initial_distribution[state]
-
         # iterate over all observations
         for frame in range(num_observations):
-            # indices for current and previous time step
-            cur = frame + 1
-            prev = frame
             # keep track of the normalisation sum
             prob_sum = 0
             # iterate over all states
             for state in range(num_states):
+                # we need to reset the current state in case this method
+                # gets called multiple times (i.e. framewise processing)
+                fwd_cur[frame, state] = 0
```

Comment:
Yes, this is a leftover from a previous code version. Back then I also initialised `fwd_cur` of length 1 during initialisation and only re-allocated it if called with observations with length > 1.